libclamav/readdb.c

```diff
@@ -312,7 +312,100 @@ cl_error_t cli_sigopts_handler(struct cli_matcher *root, const char *virname, co
     return ret;
 }
 
-#define SUB_TOKENS 4
+/**
+ * @brief Parse a regex term: a logical subsignature or yara regex string
+ *
+ * expected format => ^offset:trigger/regex/[cflags]$
+ *
+ * @param root      The matcher root (engine structure containing loaded signature patterns for matching)
+ * @param virname   Name of signature that this regex subsig came from.
+ * @param hexsig    The string containing the regex
+ * @param offset    The string offset where the pattern starts
+ * @param lsigid    An array of 2 uint32_t numbers: lsig_id and subsig_id. May be NULL for testing.
+ * @param options   Database options.  See CL_DB_* macros in clamav.h.
+ * @return cl_error_t
+ */
+static cl_error_t readdb_load_regex_subsignature(struct cli_matcher *root, const char *virname, char *hexsig,
+                                                 const char *offset, const uint32_t *lsigid, unsigned int options)
+{
+    cl_error_t status = CL_EPARSE;
+    cl_error_t ret;
+    char *hexcpy = NULL;
+    char *start  = NULL;
+    char *end    = NULL;
+
+    const char *trigger, *pattern, *cflags;
+    int subtokens_count;
+// The maximum number of `:` delimited fields in a regex subsignature.
+#define MAX_REGEX_SUB_TOKENS 4
+    char *subtokens[MAX_REGEX_SUB_TOKENS + 1];
+    const char *sig;
+
+    subtokens_count = cli_ldbtokenize(hexsig, ':', MAX_REGEX_SUB_TOKENS + 1, (const char **)subtokens, 0);
+    if (!subtokens_count) {
+        cli_errmsg("Invalid or unsupported ldb subsignature format\n");
+        status = CL_EMALFDB;
+        goto done;
+    }
+
+    if ((subtokens_count % 2) == 0)
+        offset = subtokens[0];
+
+    sig = (subtokens_count % 2) ? subtokens[0] : subtokens[1];
+
+    /* get copied */
+    hexcpy = cli_strdup(sig);
+    if (!hexcpy) {
+        status = CL_EMEM;
+        goto done;
+    }
+
+    /* get delimiters-ed */
+    start = strchr(hexcpy, '/');
```

Comment:
We discovered that unescaped slashes are tolerated within PCRE subsigs -- this would be why.

Reply:
I don't think so. This is to find the start and end of the regex string.  It doesn't have anything to do with `/` inside the regex string.  it just gets the first and last one to find the bounds.

Reply:
No worries -- this relates to some [likely inaccurate] documentation on PCRE subsignature requirements.